packages/styleguide/src/components/Editor/Core/Forms.tsx

```diff
@@ -46,35 +53,89 @@ const getForm = (
+  let parent = Editor.parent(editor, rootNode[1])
   while (
-    !elConfig[topLevelNode[0].type].attrs?.stopFormIteration &&
+    !elConfig[rootNode[0].type].attrs?.stopFormIteration &&
     SlateElement.isElement(parent[0]) &&
     !elConfig[parent[0].type].attrs?.stopFormIteration
   ) {
-    topLevelNode = parent as NodeEntry<CustomElement>
-    parent = Editor.parent(editor, topLevelNode[1])
+    rootNode = parent as NodeEntry<CustomElement>
+    parent = Editor.parent(editor, rootNode[1])
   }
+  return rootNode
+}
 
+const getDescendantForms = (
+  editor: CustomEditor,
+  rootNode: NodeEntry<CustomElement>,
+): FormData[] => {
   let forms: FormData[] = []
   for (const [n, p] of Editor.nodes(editor, {
-    at: topLevelNode[1],
+    at: rootNode[1],
   })) {
-    if (SlateElement.isElement(n) && isDescendant(topLevelNode, [n, p])) {
+    if (SlateElement.isElement(n) && isDescendant(rootNode, [n, p])) {
       const stop = elConfig[n.type].attrs?.stopFormIteration
-      const isTopLevel = p.length === topLevelNode[1].length
+      const isTopLevel = p.length === rootNode[1].length
       if (stop && !isTopLevel) break
       const currentForm = getForm(editor, [n, p])
       forms = forms.concat(currentForm)
       if (stop) break
     }
   }
+  return forms
+}
+
+const getOverlappingForm = (
+  editor: CustomEditor,
+  rootForm?: FormData,
+): FormData => {
+  if (!rootForm) return
+  const topLevelNode = rootForm.node
+  if (!elConfig[topLevelNode[0].type]?.attrs?.isInline) return
+  const [n, p] = Editor.parent(editor, topLevelNode[1])
+  if (
+    SlateElement.isElement(n) &&
+    n.children.length === 3 &&
+    Text.isText(n.children[0]) &&
+    n.children[0].text === '' &&
+    Text.isText(n.children[2]) &&
+    n.children[2].text === '' &&
+    Editor.isInline(editor, n)
+  ) {
+    return getForm(editor, [n, p])
+  }
+}
+
+// this function has to be able to handle various cases:
+// 1) forms for structured elements. E.g. for pictures, the image form should also come up
+//    when clicking on the caption, when if the form is located on a sibling branch. We
+//    also want that the various forms always show up in the same order (e.g. #1 picture size,
+//    #2 image upload) regardless where the writer clicked (top-level picture, image, or caption)
+// 2) nested inlines. E.g. a link overlapping fully with a memo. We want to bring up both forms
+//
+// To handle 1) we start by going as far as specified in the config (until we bump into
+// an element with stopIteration flag set to true). Then we go down the tree from the root
+// node recontruct our array of forms, always in the same order.
+//
+// To handle 2) we check whether the top level form element is:
+//  - inline
+//  - has inline ancestor
+//  - has inline ancestor which only contains last element
+export const getForms = (editor: CustomEditor, path: number[]): FormData[] => {
+  if (!path || path === []) return []
```

Comment:
Huch, what's the purpose of the second condition?

Reply:
Not sure if you refer to the condition in the code or the comments 😄 

`- has inline ancestor which only contains last element`
we only want to show a potential parent form if e.g. the memo fully overlaps with the link (if the memo is bigger than the link the writer can trigger the memo by clicking on the memo outside of the link. I rephrased for a bit more clarity.

`path === []`
this means the whole editor text is selected. It doesn't make sense to show forms in this case